cluster.go

```diff
@@ -311,22 +312,40 @@ func (c *Cluster) alertsHandler() {
 		case <-c.ctx.Done():
 			return
 		case alrt := <-c.monitor.Alerts():
-			// only the leader handles alerts
-			leader, err := c.consensus.Leader(c.ctx)
-			if err == nil && leader == c.id {
-				logger.Warningf(
-					"Peer %s received alert for %s in %s",
-					c.id, alrt.MetricName, alrt.Peer,
-				)
-				switch alrt.MetricName {
-				case pingMetricName:
+			list := c.state.List(c.ctx)
+			for _, pin := range list {
+				if len(pin.Allocations) < 1 && containsPeer(pin.Allocations, alrt.Peer) {
+					logger.Error("a pin with only one allocation cannot be repinned")
+					logger.Error("to make repinning possible, pin with a replication factor of 2+")
+					continue
+				}
+				if c.shouldPeerRepinCid(alrt.Peer, pin) {
 					c.repinFromPeer(c.ctx, alrt.Peer)
 				}
 			}
 		}
 	}
 }
 
+// shouldPeerRepinCid returns true if the current peer is the top of the
+// allocs list. The failed peer is ignored, i.e. if current peer is
+// second and the failed peer is first, the function will still
+// return true.
+func (c *Cluster) shouldPeerRepinCid(failed peer.ID, pin *api.Pin) bool {
+	if containsPeer(pin.Allocations, failed) && containsPeer(pin.Allocations, c.id) {
+		allocs := sort.StringSlice(api.PeersToStrings(pin.Allocations))
```

Comment:
maybe you can use https://godoc.org/github.com/libp2p/go-libp2p-peer#IDSlice so that you can sort without stringifying..